Code/Framework/AzToolsFramework/Tests/Input/QtEventToAzInputMapperTests.cpp

```diff
@@ -512,4 +516,32 @@ namespace UnitTest
             return info.param.m_az.GetName();
         }
     );
+
+    TEST_F(QtEventToAzInputMapperFixture, MouseWrapMouseViewportQtEventToAzInputMapperFixture) 
```

Comment:
update: This isn't a great name for the test, ideally we want the name to encapsulate a behavior that we're verifying.

For me a super valuable test would be, having one were we set the cursor to say, 20 units from the right side of the screen (rect), and then move it with a delta of 40 units. In Wrapping mode, the cursor should end up at 20 units from the left side of the screen, in non-wrapping mode, it should be clamped to the right most edge of the screen (and ideally we want this for X min/max and Y min/max of the rect.

In a perfect would we have separate tests for each of these, all with names describing exactly what behavior they're encoding. Any duplicate code can be moved to a new fixture or just put in helper functions we can call in each test. We ideally want tests to be short and only test one thing at a time.

If I have time today I'll try and sync this branch and create something along the lines I'm thinking of (sorry if any of this explanation is not clear, please feel free to ask my to clarify anything).

I've written up some guidance on what makes a good test I can share too that might be helpful as well

Reply:
> 

I tried that. the mouse like stops moving and is just on the left side where the value is 1. does it interrupt the mouse move if you move the cursor like that :?

Reply:
Oh interesting... ah maybe you are right and we need to move in increments of 1 unit instead of a single move... I guess in reality we're generating lots of small mouse move events so perhaps 40 iterations instead moving one unit at a time, or 20 iterations moving two units at a time and that should work I think 🤞

Reply:
```
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from QtEventToAzInputMapperFixture
[ RUN      ] QtEventToAzInputMapperFixture.MouseWrapMouseViewportQtEventToAzInputMapperFixture
/home/michaelpollind/project/o3de/Code/Framework/AzToolsFramework/Tests/Input/QtEventToAzInputMapperTests.cpp:544: Failure
The difference between endPosition.x() and expectedPosition.x() is 18, which exceeds 1.0f, where
endPosition.x() evaluates to 2,
expectedPosition.x() evaluates to 20, and
1.0f evaluates to 1.
[  FAILED  ] QtEventToAzInputMapperFixture.MouseWrapMouseViewportQtEventToAzInputMapperFixture (20 ms)
[----------] 1 test from QtEventToAzInputMapperFixture (20 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (20 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] QtEventToAzInputMapperFixture.MouseWrapMouseViewportQtEventToAzInputMapperFixture

 1 FAILED TEST
OKAY AzRunUnitTests() returned 1
Retry command: 
./AzTestRunner libAzToolsFramework.Tests.so AzRunUnitTests
Returning code: 1

```

Reply:
Is this the failure that happens when using small increments too? (sorry I'm about to go offline but I'll check back tomorrow first thing). It might be interesting to step through the code and check what it's doing or if something isn't behaving as expected. You could also so some quick sanity checks to see that the calls to `setCursorPos` do the right thing (maybe some event is not being handled in the test). This stuff is always a bit of a pain to get right but I promise we'll walk away with a much stronger understanding of exactly what is going on here and can be sure future maintainers or refactorings won't inadvertently break something. Thanks for taking the time to investigate this, I'll try grab your branch tomorrow too to take a look :+1:

Reply:
this is with just default mouse movement

Reply:
I was looking at the Qt code for mouseMove. qtestmouse.h#225 looks like it just sets the mouse cursor position. so it will go over then I set the position to the left edge. could account for the delta and add that back on the left side if we want to do that. 

```
case MouseMove:
                QCursor::setPos(widget->mapToGlobal(pos));
                qApp->processEvents();
                return;
```

Reply:
https://code.woboq.org/qt5/qtbase/src/testlib/qtestmouse.h.html#182

something seems wrong with AzToolsFrameTestHelper#mouseMove. 

mapToGlobal is getting called twice once for nextPosition and again with QTest::mouseMove

Reply:
I left a comment to explain this issue below (see `AzToolsFrameworkHelpers.h` and `MouseMove`) - hopefully you can use them instead and it should work as expected